### ISSUE TITLE
Not possible to implement custom FS due to lack of public FS API

### ISSUE THREAD

**Issue by vitonsky** [electric-sql/pglite#481]
# The problem

I trying to implement my own storage to keep data in RAM and efficient synchronization with encrypted file system.

I checked [NodeFS implementation](https://github.com/electric-sql/pglite/blob/9aff6739389647ee55e439c7b08a970c40ac3329/packages/pglite/src/fs/nodefs.ts) as reference for my own, but when i trying to implement it i have problem - some API has not been published.

For example, reference implementation have next code:

https://github.com/electric-sql/pglite/blob/9aff6739389647ee55e439c7b08a970c40ac3329/packages/pglite/src/fs/nodefs.ts#L3-L4

I trying to use `PGDATA` in my code too, but i can't found it as and some other API like `EmscriptenBuiltinFilesystem` or `BaseFilesystem`, since they are not public available.

See this example of [nodefs.d.ts](https://www.unpkg.com/@electric-sql/pglite@0.2.15/dist/fs/nodefs.d.ts) file from bundle:

![image](https://github.com/user-attachments/assets/e6b401e0-0499-480c-9cde-76d5702d2570)

Here import from chunk file with generated name, so end users can't guess this filename and entities names that are transformed too. Every new release such names may be changed and they have no types.

# The solution

My proposal is to make FS API are public available for end users.

Specifically, import all entities that may be useful for implement custom FS. Also, add a minimal docs to explain some basic concepts end users need to know to implement custom FS well.

# Use cases

My use case is i looking for an embedded DB that will keep all data in RAM, and will provide a ways to sync this data with encrypted FS that implemented by my application code.

I tried to use `dumpDataDir` + `loadDataDir` approach, but it does not work for me, since my application requires frequent synchronization for every single change in database, and in my tests this way with `dumpDataDir` + `loadDataDir` takes a seconds (2.3 seconds) that is unacceptable. The acceptable time for single sync is 10-100ms, but not more.

If you think i can improve performance with no custom FS, please tell about it. But keep in mind that we can't save any data directly at user file system, since it would be a security failure. We may interact only with virtual file system implemented by application.

**Comment by samwillis** [comment]
Hey @vitonsky 

We absolutely want to make it easy to build custom VFSs. There is an internal FS abstraction we use, I have opened a PR to export it.

This can be used to either wrap an Emscripten VFS (as we do for the IDB and Node VFSs), or to create a completely custom one like the OPFS-AHP VFS.

**This API is not stable**, we may significantly change it in future, but it's good to use for now if you are happy to make changes later. If there are changes, they will still be mostly analogous to each other. When it stabilises we will document it properly.

The best place to look for how to use it is the OPFS-AHP VFS.

PR is here: https://github.com/electric-sql/pglite/pull/493 and should be in the next release.

**Comment by vitonsky** [comment]
@samwillis hey, i've file another relevant issue about custom FS https://github.com/electric-sql/pglite/issues/810

TLDR; A custom FS cannot be implemented due to weird behaviour of API. This issue has been closed, I'm not sure why. I'm sure PGLite is unusable as embedded DB for real world apps because of this problem. Could you check it.

It seems the @tdrz have some personal negative to me, and he is not ready to discuss this problem or help me to understand how exactly I can fix this problem in PGLite myself.

He [says](https://github.com/electric-sql/pglite/issues/819#issuecomment-3915481877)
> PGlite is open source. You can implement whichever feature you want on top of your own fork.
> I'm closing this issue as it seems it is not really useful to anyone.

and closes my issues with no attention to a problem i report.

Let's keep calm and will try to think together how to fix this problem. I'm ready to contribute to PGLite, but I need some help.
